tempesta_fw/http.c

```diff
@@ -796,122 +1501,272 @@ tfw_http_adjust_resp(TfwHttpResp *resp, TfwHttpReq *req)
 				     TFW_HTTP_HDR_SERVER, 0);
 }
 
+/*
+ * Forward responses in @ret_queue to the client in correct order.
+ *
+ * In case of error the client connection must be closed immediately.
+ * Otherwise, the correct order of responses will be broken. Unsent
+ * responses are taken care of by the caller.
+ */
+static void
+__tfw_http_resp_fwd(TfwCliConn *cli_conn, struct list_head *ret_queue)
+{
+	TfwHttpReq *req, *tmp;
+
+	list_for_each_entry_safe(req, tmp, ret_queue, msg.seq_list) {
+		BUG_ON(!req->resp);
+		if (tfw_cli_conn_send(cli_conn, (TfwMsg *)req->resp)) {
+			ss_close_sync(cli_conn->sk, true);
+			return;
+		}
+		__tfw_http_resp_pair_free(req);
+		TFW_INC_STAT_BH(serv.msgs_forwarded);
+	}
+}
+
+/*
+ * Pair response @resp with request @req in @seq_queue. Then, starting
+ * with the first request in @seq_queue, pick consecutive requests that
+ * have a paired response. Move those requests to the list of returned
+ * responses @ret_queue. Sequentially send responses from @ret_queue to
+ * the client.
+ */
+void
+tfw_http_resp_fwd(TfwHttpReq *req, TfwHttpResp *resp)
+{
+	TfwCliConn *cli_conn = (TfwCliConn *)req->conn;
+	struct list_head *seq_queue = &cli_conn->seq_queue;
+	struct list_head *req_retent = NULL;
+	LIST_HEAD(ret_queue);
+
+	TFW_DBG2("%s: req=[%p], resp=[%p]\n", __func__, req, resp);
+
+	/*
+	 * If the list is empty, then it's either a bug, or the client
+	 * connection had been closed. If it's a bug, then the correct
+	 * order of responses to requests may be broken. The connection
+	 * with the client must to be closed immediately.
+	 *
+	 * Doing ss_close_sync() on client connection's socket is safe
+	 * as long as @req that holds a reference to the connection is
+	 * not freed.
+	 */
+	spin_lock(&cli_conn->seq_qlock);
+	if (unlikely(list_empty(seq_queue))) {
+		spin_unlock(&cli_conn->seq_qlock);
+		TFW_DBG2("%s: The client's request missing: conn=[%p]\n",
+			 __func__, cli_conn);
+		ss_close_sync(cli_conn->sk, true);
```

Comment:
There's nothing wrong with queueing. However, without the SYNC flag queueing is NOT guaranteed. We have to close the connection at this point as the pairing may get broken.

With `ss_close_sync()` the connection may get closed either right away, or be queued for a later execution. Either way, closing is guaranteed. If it's queued, then scheduled earlier responses will be sent out before the connection is closed, but that's will not happen if the connection is closed right away in the same thread.

Note that this code may (and does, in fact) get executed multiple times, which causes multiple calls to close the connection. That's why there's a check for a duplicate code in `ss_send()` in the code path that gets executed immediately in the function's calling thread. Is there a better way?